lib/_stream_writable.js

```diff
@@ -547,67 +504,52 @@ function afterWrite(stream, state, count, cb) {
+  const buffered = state.buffered;
+  const bufferedIndex = state.bufferedIndex;
+  const bufferedLength = buffered.length - bufferedIndex;
+
+  if (!bufferedLength || state.corked || state.bufferProcessing) {
+    return;
+  }
+
   state.bufferProcessing = true;
-  var entry = state.bufferedRequest;
-
-  if (stream._writev && entry && entry.next) {
-    // Fast case, write everything using _writev()
-    var l = state.bufferedRequestCount;
-    var buffer = new Array(l);
-    var holder = state.corkedRequestsFree;
-    holder.entry = entry;
-
-    var count = 0;
-    var allBuffers = true;
-    while (entry) {
-      buffer[count] = entry;
-      if (!entry.isBuf)
+  if (bufferedLength > 1 && stream._writev) {
+    state.pendingcb++;
+
+    const buffer = new Array(bufferedLength);
+    let allBuffers = true;
+    for (let n = bufferedIndex; n < buffered.length; ++n) {
+      buffer[n] = buffered[n];
+      if (allBuffers && buffered[n].encoding !== 'buffer') {
         allBuffers = false;
-      entry = entry.next;
-      count += 1;
+      }
     }
     buffer.allBuffers = allBuffers;
 
-    doWrite(stream, state, true, state.length, buffer, '', holder.finish);
+    doWrite(stream, state, true, state.length, buffer, '', (err) => {
+      for (let n = bufferedIndex; n < buffered.length; ++n) {
+        state.pendingcb--;
+        buffered[n].callback(err);
+      }
+    });
 
-    // doWrite is almost always async, defer these to save a bit of time
-    // as the hot path ends with doWrite
-    state.pendingcb++;
-    state.lastBufferedRequest = null;
-    if (holder.next) {
-      state.corkedRequestsFree = holder.next;
-      holder.next = null;
-    } else {
-      var corkReq = { next: null, entry: null, finish: undefined };
-      corkReq.finish = onCorkedFinish.bind(undefined, corkReq, state);
-      state.corkedRequestsFree = corkReq;
-    }
-    state.bufferedRequestCount = 0;
+    state.buffered = [];
+    state.bufferedIndex = 0;
   } else {
-    // Slow case, write chunks one-by-one
-    while (entry) {
-      var chunk = entry.chunk;
-      var encoding = entry.encoding;
-      var cb = entry.callback;
-      var len = state.objectMode ? 1 : chunk.length;
-
-      doWrite(stream, state, false, len, chunk, encoding, cb);
-      entry = entry.next;
-      state.bufferedRequestCount--;
-      // If we didn't call the onwrite immediately, then
-      // it means that we need to wait until it does.
-      // also, that means that the chunk and cb are currently
-      // being processed, so move the buffer counter past them.
-      if (state.writing) {
-        break;
-      }
+    let i = bufferedIndex;
+    do {
+      const { chunk, encoding, callback } = buffered[i++];
+      const len = state.objectMode ? 1 : chunk.length;
+      doWrite(stream, state, false, len, chunk, encoding, callback);
+    } while (i < buffered.length && !state.writing);
+
+    if (i > 256 || i === buffered.length) {
+      buffered.splice(0, i);
```

Comment:
This essentially means that `Buffer`s could be held onto indefinitely if the subsequent writes here take a long time to finish, right?

Reply:
Nice catch. Fixed.